Implementing the storage version API, Part I

caesarxuchao opened the pull request:
/sig-apimachinery
/assign @deads2k @yliaog @lavalamp 
/kind feature
[KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md)

This PR implements:
1. the storage version API.
2. Let the kube-aggregator host the storage version API.
3. kube-apiserver updates the storage version during bootstrap.
4. A filter that blocks certain requests during kube-apiserver bootstrap until the kube-apiserver has updated the storage versions.
5. An integration test that shows the filer works, and that the apiserver works properly after the storage version has been updated.

This PR is already huge, so I'll do the remaining tasks in follow-up PRs. Some major follow-ups are:
1. the filter should block write requests to "apiservices" and "crd".
2. the garbage collection of obsolete storage version objects.
3. a test showing that the API works fine in HA cluster.
4. custom resource handler needs to update the storage version the CR before accepting write requests.

caesarxuchao commented:
/retest

caesarxuchao commented:
/retest

caesarxuchao commented:
/retest

Commit:
scripts for code generation

Commit:
Add storage version API.

KEP is
https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md.

Commit:
generated client and other generated code changes.

Commit:
Make it possible to include group name in the etcd path for resources
installed with the DefaultStorageFactory.

Commit:
Add a feature flag

Commit:
cluster role for apiserver to update storage version

Commit:
make some rbac and scheduling post start hooks tolerate the apiserver
bootstrap delay caused by installing storage versions.

Commit:
golint exceptions

Commit:
test-only, start local cluster with storage version enabled

caesarxuchao commented:
/unassign

caesarxuchao commented:
/retest

caesarxuchao commented:
cc @sanchezl

caesarxuchao commented:
/kind api-review

caesarxuchao commented:
/api-review

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:
```diff
@@ -0,0 +1,112 @@
+	}
+	var completed bool
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		fmt.Println("CHAO: in filter")
```
debug line

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:
```diff
@@ -0,0 +1,112 @@
+func (svb *StorageVersionBarrier) Completed() bool {
+	svb.mu.RLock()
+	defer svb.mu.RUnlock()
+	return svb.completed
```
frequently read, rarely written, I tend to think atomic.Value

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:
```diff
@@ -0,0 +1,112 @@
+	var completed bool
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		fmt.Println("CHAO: in filter")
+		if completed {
```
pretty sure that the `Handle` method is called on multiple threads, making this variable access a race

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:
```diff
@@ -0,0 +1,112 @@
+
+// StorageVersionBarrier indicates if the aggregator, kube-apiserver, and the
+// apiextensions-apiserver have completed reporting their storage versions.
+type StorageVersionBarrier struct {
```
if we made this hold all the StorageVersions that need to be registered, then we could selectively hold only requests for resources that don't yet have the storage versions registered.  Also, this could contain a controller fired by post-starthook that waits for a runonce to give a best shot at registration before becoming ready (tries once).  And I think it would simplify the registration process by allowing the installAPI to fully contain the detail of setting this to reconcile on its behalf.

Commit:
Add an integration test.

To make sure that the storage version filter can block certain requests until
the storage version updates are completed, and that the apiserver works
properly after the storage version updates are done.

Commit:
Let kube-aggregator host the storage version API

Commit:
Collect storage versions as ResourceInfo when installing API endpoints.

Commit:
Add a generic filter that blocks certain write requests before
StorageVersions are updated during apiserver bootstrap.

Also add a poststarthook to the aggregator which updates the
StorageVersions via the storageversion.Manager

Commit:
Add the storageversion.Manager interface

The interface exposes methods to record storage versions in
ResourceInfo, to update StorageVersions, and to check if storageVersion
updated have completed.

Commit:
generated

caesarxuchao commented:
/retest

caesarxuchao commented:
The pull-kubernetes-e2e-gce-alpha-features failed because the alpha API (storageversion) was not enabled.

caesarxuchao commented:
To fix pull-kubernetes-e2e-gce-alpha-features, we need to change its job yaml. It only runs some [alpha e2e tests](https://github.com/kubernetes/test-infra/blame/17b0b8c63acd8126c1a54874b46f4f3eea4f77a4/config/jobs/kubernetes/sig-cloud-provider/gcp/gcp-gce.yaml#L292).

OTOH, the pull-kubernetes-local-e2e ran all e2e conformance tests. It passed with the storage version feature enabled. So I don't plan to fix pull-kubernetes-e2e-gce-alpha-features.

deads2k commented on staging/src/k8s.io/apiserver/pkg/storageversion/manager.go:
```diff
@@ -0,0 +1,194 @@
+	}
+}
+
+func (s *DefaultManager) addGroupResource(group, resource string) {
```
prefer schema.GroupResource

deads2k commented on staging/src/k8s.io/apiserver/pkg/storageversion/manager.go:
```diff
@@ -0,0 +1,194 @@
+			return
+		}
+		klog.V(2).Infof("successfully updated storage version for %v", r.Resource.Name)
+		s.RemoveResourceInfo(r)
```
I suspect we will want to know the full list of ones we handled.  Instead, can you mark it as completed.

deads2k commented on staging/src/k8s.io/apiserver/pkg/storageversion/manager.go:
```diff
@@ -0,0 +1,194 @@
+
+	mu             sync.RWMutex
+	resources      map[*ResourceInfo]struct{}
+	groupResources map[string]map[string]struct{}
```
How about `groupResources map[schema.GroupResource]resourceStatus`

```go
type resourceStatus struct{
    done bool
    lastErr error
}
```

for our future inspection.

deads2k replied:
This would also eliminate the parallel maps

deads2k commented on staging/src/k8s.io/apiserver/pkg/storageversion/manager.go:
```diff
@@ -0,0 +1,194 @@
+
+// UpdateStorageVersions updates the StorageVersions. If the updates are
+// successful, following calls to Completed() returns true.
+func (s *DefaultManager) UpdateStorageVersions(loopbackClientConfig *rest.Config, serverID string) {
```
`kubeAPIServerClientConfig` .  On aggregated APIs, loopback means a different thing.

deads2k commented on staging/src/k8s.io/apiserver/pkg/storageversion/manager.go:
```diff
@@ -0,0 +1,194 @@
+	for r := range resources {
+		r.DecodableVersions = decodableVersions(r.EquivalentResourceMapper, r.Group, r.Resource.Name)
+		if err := updateStorageVersionFor(sc, serverID, r.Group+"."+r.Resource.Name, r.EncodingVersion, r.DecodableVersions); err != nil {
+			klog.Fatalf("failed to update storage version for %v", r.Resource.Name)
```
not Fatal.  perhaps `utilruntime.HandleError`

deads2k commented on staging/src/k8s.io/apiserver/pkg/storageversion/updater.go:
```diff
@@ -0,0 +1,116 @@
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	if err != nil && apierrors.IsNotFound(err) {
```
`err != nil` isn't needed

deads2k commented on staging/src/k8s.io/apiserver/pkg/storageversion/updater.go:
```diff
@@ -0,0 +1,116 @@
+	return err
+}
+
+func localUpdateStorageVersion(sv *v1alpha1.StorageVersion, apiserverID, encodingVersion string, decodableVersions []string) {
```
please refactor to build a new object to return that can be assigned instead of mutating the input.

deads2k commented on staging/src/k8s.io/apiserver/pkg/apis/apiserverinternal/v1alpha1/types.go:
```diff
@@ -0,0 +1,119 @@
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Spec is omitted because there is no spec field.
+
```
stub in an empty spec please, so the proto looks normal and the object doesn't look too weird.

deads2k commented:
@caesarxuchao do you have a separate PR with just the API?

deads2k commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -412,11 +420,15 @@ func (s *GenericAPIServer) installAPIResources(apiPrefix string, apiGroupInfo *A
+		resources = append(resources, r...)
 	}
 
+	s.StorageVersion.AddResourceInfo(resources...)
```
Does this happen before we start listening?  if so, add a comment that this approach is safe because of this.

deads2k commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:
```diff
@@ -192,8 +196,19 @@ func (c completedConfig) NewWithDelegate(delegationTarget genericapiserver.Deleg
-	apiGroupInfo := apiservicerest.NewRESTStorage(c.GenericConfig.MergedResourceConfig, c.GenericConfig.RESTOptionsGetter)
-	if err := s.GenericAPIServer.InstallAPIGroup(&apiGroupInfo); err != nil {
+	rp := apiserverinternalrest.StorageProvider{}
+	apiGroupInfo, err := rp.NewRESTStorage(c.GenericConfig.MergedResourceConfig, c.GenericConfig.RESTOptionsGetter)
```
I expected to see this in the kube-apiserver, not in the aggregator.

caesarxuchao replied:
The reason I let the aggregator host it is because the storage version API contains information about all apiservers (kube-apiserver, apiextension apiserver, aggregator apiserver, and custom apiservers). It's analogous to APIService, which is also hosted by the aggrgator. WDYT?

deads2k replied:
> The reason I let the aggregator host it is because the storage version API contains information about all apiservers (kube-apiserver, apiextension apiserver, aggregator apiserver, and custom apiservers). It's analogous to APIService, which is also hosted by the aggrgator. WDYT?

I think this is closer to the idea of webhook admission configuration.  It's provided by the kube-apiserver for consumption and usage by other servers, but aggregation is focused on its single mission

caesarxuchao commented:
Inherited by #92064 and #92459.
